### ISSUE TITLE
Pre-built binary for openvpn3 client for the new stable release of Debian (version 12 codename bookworm)

### ISSUE THREAD

**Issue by jim-barber-he** [OpenVPN/openvpn3-linux#193]
The latest Debian stable release came out on the 10th of June 2023.
Currently there is no pre-built openvpn3 Debian package built for the `bookworm` distribution.
This is a request for adding a new package to support this since the other .deb packages have dependencies on packages that are too old.

**Comment by mschlachter** [comment]
I contacted support about this and they told me it's planned for the v21 release. I would love to have a timeline for when this will be available if possible.

**Comment by dsommers** [comment]
This is not a promise.  But I do hope I can manage to have a v21 release out by end of August/early September.  This will definitely include Debian 12 builds.  It is holiday season now, so not much will happen in July.

**Comment by samiramer** [comment]
Hey @dsommers do you have an update on the release timeline?  This would be a lifesaver.

**Comment by samiramer** [comment]
Related to #201 and #171

**Comment by dsommers** [comment]
The v21 release has been delayed due to work related to solve #171 as well as some issues reported in the Core library by our internal QA.  Hopefully a v21 for stable distribution releases (aka LTS distros, Debian, Ubuntu LTS, RHEL/CentOS) will be ready for internal QA within 2 weeks.  Since #171 has now been taken out of the v21 release, I'm already wrapping up the last outstanding issues to get a proper build and packaging done.  Once QA has approved the release, it will be shipped.

The v21 release will ship with an updated [OpenVPN 3 Core library](https://github.com/OpenVPN/openvpn3), which is currently going through QA now.  Even though issues found by QA in this library has so far had a very small impact on OpenVPN 3 Linux (it is also used in OpenVPN Connect), there are a still a few potential issues which might affect this project as well.

**Comment by bhack** [comment]
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=904044

**Comment by Drizzt321** [comment]
@dsommers any further work on v21? I can muddle through with old openvpn2 client, but the v3 would be great to be able to use.

**Comment by dsommers** [comment]
@Drizzt321 The v21 is going through QA these days.  If all goes well, end of the month is the main goal for the final release.

I'm pushing out fixes needed for this release in the [releaseprep/v21 branch](https://github.com/OpenVPN/openvpn3-linux/commits/releaseprep/v21).

**Comment by dsommers** [comment]
@Drizzt321 I've attached here the latest QA build for Debian 12.  Feel free to test it if you want.  Any feedback would be valuable.

[openvpn3_0-20230919git6361bd5-1+bookworm_amd64.deb.gz](https://github.com/OpenVPN/openvpn3-linux/files/12671852/openvpn3_0-20230919git6361bd5-1%2Bbookworm_amd64.deb.gz)

(Since GitHub is picky about such attachments, you need to `gunzip` this file before running:

     apt install ./openvpn3_0-20230919git6361bd5-1+bookworm_amd64.deb

**Comment by bhack** [comment]
@mleeman is also preparing a regular/official Debian package at:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=904044#27

**Comment by mleeman** [comment]
Dang, busted :-P 

I've added some comments in the ticket, the highlights are:

1. The debian builders do not like git submodules (no network access). In order to avoid having to recompress the releases; I decided to make 2 packages: openvpn3 (library) and openvpn3-linux. The package names match the repository names; this allows easy maintenance with git-buildpackage when a  new release is tagged
2. I have added pkg-config support in openvpn3 for better support in the client compilation.
3. Minor patches are applied to avoid git and git submodule checkout/update
4. I need to add a binary conflict to openvpn3 to avoid problems with ppl that have both repos enabled

Feel free to comment, review
https://salsa.debian.org/televic-team/openvpn3
https://salsa.debian.org/televic-team/openvpn3-linux

**Comment by dsommers** [comment]
@mleeman That's really cool you're looking into this!

In regards to i386 ... There are some issues on some of the C++ templates converting between C++ data types to D-Bus data types which is not really happy on 32-bit platforms.  Since the wast majority of users are on 64-bit platforms, this has not been a high priority to resolve.

The documentation need to be updated also, the minimum C++ standard now will be C++17; this is due to the OpenVPN 3 Core library moved to that a while back.  Further, the glib2 refactoring for #171 will also require C++17.

In regards to source code, we will always publish source code as tarballs as well; which includes everything needed to build with `./configure`.  The v20 is pushed out here:
https://swupdate.openvpn.net/community/releases/openvpn3-linux-20.tar.xz
https://swupdate.openvpn.net/community/releases/openvpn3-linux-20.tar.xz.asc

If you want, I can provide a similar development source tarball for the binary posted a bit earlier today.

Please reach out to me if you see any changes on our end to make the packaging better and/or smoother.  Ideally we should not need to ship our own builds as a third-party repos.  And this goes both for potential packaging efforts of the OpenVPN 3 Core library as well as the OpenVPN 3 Linux project.

In regards to packaging, lintian will complain about D-Bus policies being too broad.  I discussed a missing feature with the upstream "dbus-daemon" developers some years ago, and got acceptance for the `send_destination_prefix` support.  IIRC, that arrived in dbus-daemon-1.14, but I'm uncertain about dbus-broker support.  This is basically the main detail to get rid of the lintian complaints.

**Comment by mleeman** [comment]
To me, it does not really matter; packaging follows (as much as possible) upstream.

I have mirrored the packaging  structure on the git repo; so if you tag your release in github, I can fetch the tarball from there.

`
$ cat debian/watch 
version=4
opts=filenamemangle=s/.+\/v?(\d\S+)\.tar\.gz/openvpn3-$1\.tar\.gz/ \
  https://github.com/openvpn/openvpn3/tags .*/v?(\d\S+)@ARCHIVE_EXT@
`

If you prefer to have full/integrated tarballs that you release on your website, I can change the watch files to match.

I have indeed added a number of lintian-overrides (needs to be reviewed over time), but we have a working solution atm.

As for patches,  I'll send the ones that are valuable to the mailing list once the dust settles.

**Comment by mleeman** [comment]
FYI, there is some low hanging fruit (spelling errors) that got picked up by the debian tooling

https://mentors.debian.net/package/openvpn3-linux/
https://mentors.debian.net/package/openvpn3/


```
    authentification authentication [usr/bin/ovpncli]
    authentification authentication [usr/lib/x86_64-linux-gnu/_ovpncli.so]
    laoding loading [usr/bin/ovpncli]
    laoding loading [usr/lib/x86_64-linux-gnu/_ovpncli.so]
```

and 

```
    Adress Address [usr/libexec/openvpn3-linux/openvpn3-service-netcfg]
    Transfered Transferred [usr/libexec/openvpn3-linux/openvpn3-service-configmgr]
    Transfered Transferred [usr/libexec/openvpn3-linux/openvpn3-service-sessionmgr]
    Unkown Unknown [usr/sbin/openvpn3-admin]
    authentification authentication [usr/bin/openvpn3]
    authentification authentication [usr/libexec/openvpn3-linux/openvpn3-service-client]
    laoding loading [usr/bin/openvpn3]
    laoding loading [usr/libexec/openvpn3-linux/openvpn3-service-aws]
    laoding loading [usr/libexec/openvpn3-linux/openvpn3-service-client]
    occured occurred [usr/sbin/openvpn3-admin]
    subsciber subscriber [usr/bin/openvpn3]
    subsciber subscriber [usr/sbin/openvpn3-admin]
    unkown unknown [usr/libexec/openvpn3-linux/openvpn3-service-client]
```

**Comment by dsommers** [comment]
* **Source tarball**
  While I see the simplicity of pulling the tarball from github, it is not a proper distribution tarball; it's a git snapshot not including any dependencies.  The tarballs we publish will have all dependencies and those tarballs are what is being used across all our builds - including Fedora Copr builds (and later upstream Fedora packaging). These tarballs are the one used through our QA processes and should also be prepared to comply to reproducible builds requirements; that the built binary will be identical - as long as the distribution and compiler chain is the same.  And the authenticity of these tarballs can be verified through our PGP signature.

* **Packaging of OpenVPN 3 Core library is dubious**
   It is primarily a source header-only library.  While it can produce a libopenvpn3.so in some cases; that is only used for clients not written in C++ where a glue layer is needed - and it depends on `swig` to work.  OpenVPN 3 Linux does not depend on a shared library (and probably won't go that path).  Packaging OpenVPN 3 Core library should probably just be a -dev package with only the header files from `./client`, `./doc` and `./openvpn`.

* I'll look at all the typos and such; that should be fine to get fixed.

**Comment by dsommers** [comment]
@mleeman Two quick patches fixing the openvpn3-linux typ0s ...

* https://codeberg.org/OpenVPN/openvpn3-linux/commit/2755c95728cae3bfe347c4153d6d4f959d02a4ca
* https://codeberg.org/OpenVPN/openvpn3-linux/commit/8881bf35c835c7deab439c87a7efc09444caef84

These are queued up for the coming v21 release.

**Comment by mleeman** [comment]
fine with me, I will adjust the client to use your published snapshots.

**Comment by dsommers** [comment]
@mleeman Good point on the publicity aspects of new releases. We will look into how to do that better.  Perhaps a quick solution for now would be to have use this file I just now pushed out: https://swupdate.openvpn.net/community/releases/openvpn3-linux.latest

If you want a different layout here, we can probably script that a bit better.

The version scheme is also fairly simple.  I'm going to use three "release categories":

* stable releases:  `v20` - `openvpn3-linux-20.tar.xz{,.asc}`
* beta releases: `v19_beta` - `openvpn3-linux-19_beta.tar.xz{,.asc}`
* development releases: `v22_dev` - `openvpn3-linux-22_dev.tar.xz{,.asc}`

Stable releases will be the only releases which might have a minor version, like `v20.1`.  This is only happening where the next stable release is "too far away" and there is an important bug or security fix needed.  In this case, the version strings will be:

* stable hotfix: `v20.1` -  `openvpn3-linux-20.1.tar.xz{,.asc}`

This `openvpn3-linux.latest` could perhaps be extended to have a "release category" tag as well.  Like `stable:20` and `development:22_dev`.  With a "stable hotfix", the `stable:`  line would be `stable:20.1`.

In a Debian and Ubuntu LTS context, I think it would be best stay on the `stable` releases in the longer run.  For Ubuntu non-LTS, the `beta` releases might be applicable.  The `development`  builds might be best being a separate repository we update more frequently.

Let me know what you think about this.

**Comment by mleeman** [comment]
- Starting from your snapshot and by removing all the patches gives a clean compilation, so your release tarball is an excellent starting point. If you prefer this route, then this is the one we'll be taking. It's easy and it's clean.
- I agree that only the stable releases should come in Debian, it's fine if someone uses the build infrastructure to experiment, but that is outside of the scope of Debian itself.
- One thing I was thinking about over the weekend is the name of the package. At the moment it is `openvpn3` or `openvpn3-linux`. When comparing this to the openvpn2 releases, a better name might be `openvpn3-client`, since (AFAIK) it only contains the client part.

Let me know what you prefer and I'll reset my salsa branches to your  to the release tarball instead of the one generated from github.

**Comment by dsommers** [comment]
Regarding 1), that sounds like the best approach to me too; I generally would like package maintainers to add as few additional patches  as possible - get them upstream first.  (I'm also a Fedora package maintainer, so I have some packaging experience as well)

We are aligned on the "stable" versions in Debian, so nothing more to discuss there.

Packaging names ... okay, so this is "complicated".  But it is related to some project names:

*  "OpenVPN 3 Linux" (`openvpn3-linux`) is this project; the (primarily, but not necessarily restricted to) Linux implementation of the "OpenVPN 3 generation"
 * "OpenVPN 3 Core Library" (`openvpn3`) is the implementation of the OpenVPN protocol as the 3rd generation.

Currently the `openvpn3-linux` project has only implemented the client side - but I would like to have a server side implementation too in the future.  In the Fedora packaging, there are more sub-packages:

* **`openvpn3`**
      Provides the **generic components**; configuration manager (`openvpn3-service-configmgr`), log service (`openvpn3-service-logger`), network configuration (`openvpn3-service-netcfg`), the Python module [1], the `openvpn3`and `openvpn3-admin` CLI tools, the `openvpn3-autoload` (deprecated, but will be shipped at least until next summer - RHEL 7 EOL), as well as some support files for all of this and some docs.
* **`openvpn3-client`**
      Provides the **client side components**: The client services (`openvpn3-service-backendstart`, `openvpn3-service-client`), the `openvpn2` and `openvpn3-as` CLI tools and the `openvpn3-systemd` helper and the `openvpn3-session@.service` systemd unit, plus other related support files, man pages and such
* **`openvpn3-addon-aws`** -
       Contains an **optional add-on component** (for AWS-VPC integration), provides the `openvpn3-service-aws` and related support files and man page
* **`openvpn3-selinux`**
       SELinux policy to confine the some of the services and allow D-Bus daemon to pass access to the tun/dco interfaces between two D-Bus enabled processes.
* **`openvpn3-devel`**
       Contains some generated files which is installed under `/usr/include/openvpn3` ... and they are not related to the Core library itself.

The intension behind these package names is that
* `openvpn3` itself is what is needed to run either a client or a server functionality.  It does not relate to the OpenVPN 3 Core library (header) files, as that is not an end-user functionality.
* `openvpn3-client` need to depend on `openvpn3`, since that component builds on the `openvpn3` basic functionality to provide the client functionality

If we would package the OpenVPN 3 Core library, that should be named `libopenvpn3-devel` or something like that.  It does not provide an end-user product, but a library to implement the OpenVPN protocol.  This code does contain a reference test client, but that is not something you would put into production as is.

The reason the Debian packaging does not have such splits as the Fedora is that I heard from a Debian maintainer many years ago that Debian generally prefers to avoid sub-packages and only use them when strictly needed.  In the Fedora land, sub-packages are more often used to split out functionality which does not need to be installed to have something working.  The `-selinux` package is also more a common way how to clearly indicate that this project will touch the SELinux policy on the system.

But you are of course free to do package this project in Debian as you see fit best with the Debian packaging guidelines.

**Comment by mleeman** [comment]
Debian packaging (in general) does tend to use smaller sub packages. What I would propose is to:

1. start by packaging `openvpn3-linux` into a package `openvpn3-client`. It describes clearly what it is: a client for openvpn3 functionality. For starters, I'd bundle all the tools that are shipped with `openvpn3-linux`.
2. it's great that there is a server component coming. At that point a bit more work will be needed; though I think that the starting point will probably not be the tarball we discussed before; but a number of different sub releases. When this is the case, we'll probably have packages with names like: `libopenvpn3-dev` (including the version to avoid conflicting with `/usr/include/openvpn`) and `libopenvpn3` that are dependencies for `openvpn3-client` and `openvpn3-server`. To some extent, this was they way I started working; but in retrospect (and until there is a server component,) it is probably overkill.

This way, the packaging names match to some extent the ones you defined for rpm packaging. As the need arises and if it is useful, I can split off components into different packages. At the moment, I would focus on getting `openvpn3` into Debian/Ubuntu as is, while staying very close to your releases: it give you a much better level of control on what you want to release and in what form.

How does this sound?

**Comment by dsommers** [comment]
That sounds great!  Currently the packages (and docs) in [our provided releases](https://community.openvpn.net/openvpn/wiki/OpenVPN3Linux) uses the `openvpn3` package name.  There are also some docs for CloudConnexa which also needs to be updated.

Perhaps we should move our (OpenVPN Inc provided) packaging to also use the `openvpn3-client` with a "Provides" `openvpn3` for a short time as we move forward?   It would be great if the packages we provide will be compliant with the upstream Debian ones and not causing conflicts in any way.

**Comment by flichtenheld** [comment]
In openvpn2 we use the official Debian/Ubuntu packages as base and update from that as needed for the Community releases. Once there is official Debian packaging for openvpn3 we should probably go to a similar model.

**Comment by mleeman** [comment]
I have an initial version, most of the code could be re-used
```
https://salsa.debian.org/televic-team/openvpn3-client
```

One lintian error popped up that will probably require me to re-package the code:

```
E: openvpn3-client source: license-problem-convert-utf-code [openvpn3-core/openvpn/common/unicode-impl.hpp]
N: 
N:   The following file source files include material under a non-free license from Unicode Inc. Therefore, it is not possible to ship this in main or contrib.
N:   
N:   This license does not grant any permission to modify the files (thus failing DFSG#3). Moreover, the license grant seems to attempt to restrict use to "products supporting the Unicode
N:   Standard" (thus failing DFSG#6).
N:   
N:   In this case a solution is to use libicu and to remove this code by repacking.
N: 
N:   Please refer to Bug#823100 for details.
N: 
N:   Visibility: error
N:   Show-Always: no
N:   Check: cruft
N: 
N:
```

Another one I need to look into, seems to be some data that is used during a test

```
E: openvpn3-client source: source-is-missing [openvpn3-core/test/unittests/comp-testdata/sum]
```

**Comment by dsommers** [comment]
We need to come up with a solution for that license issue in the OpenVPN 3 Core library.  It's essentially a public domain license, which shouldn't really be a problem - but legalese is always annoying.

that `comp-testdata/sum` is essentially just a random binary which is only used for compression testing in the Core library unit tests.  That file should also be replaced by some real random binary data and not a binary executable (which in this case is even for the Sparc CPU arch).

Both of these issues are coming from the OpenVPN 3 Core library.

**Comment by mleeman** [comment]
I am searching a bit and have found e.g.

```
https://llvm.org/doxygen/ConvertUTF_8h_source.html
```

it seems to be the same source and llvm is in `main`. The warning states:

```
N:   This license does not grant any permission to modify the files (thus failing DFSG#3). Moreover, the license grant seems to attempt to restrict use to "products supporting the Unicode
N:   Standard" (thus failing DFSG#6).
```

while the header file says (copy):

``` 
17  * Unicode, Inc. hereby grants the right to freely use the information
 18  * supplied in this file in the creation of products supporting the
 19  * Unicode Standard, and to make copies of this file in any form
 20  * for internal or external distribution as long as this notice
 21  * remains attached.
```

The file in `llvm` has a more recent copyright:

```
/*===--- ConvertUTF.h - Universal Character Names conversions ---------------===
 *
 * Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 * See https://llvm.org/LICENSE.txt for license information.
 * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 *
 *==------------------------------------------------------------------------==*/
/*
 * Copyright © 1991-2015 Unicode, Inc. All rights reserved.
 * Distributed under the Terms of Use in
 * http://www.unicode.org/copyright.html.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of the Unicode data files and any associated documentation
 * (the "Data Files") or Unicode software and any associated documentation
 * (the "Software") to deal in the Data Files or Software
 * without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, and/or sell copies of
 * the Data Files or Software, and to permit persons to whom the Data Files
 * or Software are furnished to do so, provided that
 * (a) this copyright and permission notice appear with all copies
 * of the Data Files or Software,
 * (b) this copyright and permission notice appear in associated
 * documentation, and
 * (c) there is clear notice in each modified Data File or in the Software
 * as well as in the documentation associated with the Data File(s) or
 * Software that the data or software has been modified.
 *
 * THE DATA FILES AND SOFTWARE ARE PROVIDED "AS IS", WITHOUT WARRANTY OF
 * ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT OF THIRD PARTY RIGHTS.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR HOLDERS INCLUDED IN THIS
 * NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL INDIRECT OR CONSEQUENTIAL
 * DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THE DATA FILES OR SOFTWARE.
 *
 * Except as contained in this notice, the name of a copyright holder
 * shall not be used in advertising or otherwise to promote the sale,
 * use or other dealings in these Data Files or Software without prior
 * written authorization of the copyright holder.
 */
```

Maybe there has been an update since the 2001-2004 version that is in the core lib?

**Comment by dsommers** [comment]
Good find!  Thanks a lot for your research!  Will dig into this and find a better licensed version.

This is anyhow something needed to be tackled in the [openvpn3 repository](https://github.com/OpenVPN/openvpn3/) and will need to come in the next OpenVPN 3 Core release before pulling it into this (openvpn3-linux) project.

**Comment by bhack** [comment]
A bit unrelated:
 do you think that this package is going to conflict with network manager openvpn in  distros? 
See https://github.com/OpenVPN/openvpn3-linux/issues/46

**Comment by darkbasic** [comment]
I personally hope that NetworkManager won't use openvpn3 because it doesn't support `fragment`, which is mandatory with certain providers that tunnel traffic thus breaking PMTUD: https://forum.fibra.click/d/44630-lista-nera-dei-provider-che-multiplexano-le-sessioni-ppp-via-tunnel-l2tp

**Comment by dsommers** [comment]
> A bit unrelated: do you think that this package is going to conflict with network manager openvpn in distros? See #46

@bhack No, it will not conflict with NetworkManager.  That's a different set of package names with different package dependencies.  A NetworkManager-openvpn**3** add-on is wanted and requested by many, and that will use a very different integration than the approach the current one does.

**Comment by Drizzt321** [comment]
Just getting back to this ticket, looks like a lot of good work has been done so far! :)

Looking forward to the official Debian packages, thanks for all the hard work, and *crosses fingers* that NM plugin sooner than later, even though the gnome ticket doesn't seem to have any movement recently.

**Comment by mleeman** [comment]
I was about to ask, is the update of the unicode code going to be part of 21 or not?

If not, I will continue to investigate creating a dfsg package.

At the moment, I am waiting in order to keep the official release tarball.

Just let me know what you decide and I'll continue.

**Comment by dsommers** [comment]
@mleeman The unicode change need to happen in the OpenVPN 3 Core Library first.  If that can go into a OpenVPN 3 Core v3.8.3 or if it first happens in Core v3.9 is not decided yet.  Once that is released, that's when I can start thinking about pulling in a new Core library in OpenVPN 3 Linux.

The v21 has almost completed QA.  There is one issue I need to double check on Ubuntu 20.04 - but except of that, it's looking quite good.  That release will hopefully make it out the door this week.  If you want to test a source tarball on the current QA build, let me know and I'll upload one for you.

**Comment by mleeman** [comment]
No, I will need to do a dfsg package, if it is the 20 or 22, I don't think it will matter much.

I'll need to replace that file with.

**Comment by mleeman** [comment]
No, I will need to do a dfsg package, if it is the 20 or 22, I don't think it will matter much.

I'll need to replace that file with.

update: since I seem to bump into more references to this file; I am asking around; either we need to change this and certainly a couple of other projects are wrong or this is not a problem and lintian is wrong.

**Comment by mleeman** [comment]
Feedback from debian-mentors on IRC (legalese is tricky):

```
TinoDidriksen
It is functionally the same MIT-esque license and otherwise is clearly a newer version of the file, both from the Unicode org. I'd say there is no problem and that it is a false positive. Ignore it.
```

**Comment by dsommers** [comment]
@mleeman Source tarballs are available

https://swupdate.openvpn.net/community/releases/openvpn3-linux-21.tar.xz
https://swupdate.openvpn.net/community/releases/openvpn3-linux-21.tar.xz.asc

**Comment by mleeman** [comment]
@dsommers integrated the new release on our internal distro for testing, builds fine for both bullseye and bookworm on mini-buildd, 

I'll push an update for unstable next.

**Comment by dsommers** [comment]
That's awesome work, @mleeman!  Once your work is ready and widely available through native Debian repos ... lets update the install instructions here: https://community.openvpn.net/openvpn/wiki/OpenVPN3Linux (adding a "native Debian repo" section).

**Comment by Drizzt321** [comment]
@mleeman sounds like that means it'll be in debian `unstable` repo shortly? That's great news. Sadly might mean a while before it backs it to bookworm-backports or 12.3 (if there is one) of bookworm, but I can hope. Unless I have a misunderstanding of debian repo progressions.

Although I have been considering just moving to `testing` so I get updates quicker.

**Comment by mleeman** [comment]
It should be in unstable soon, and in testing (Trixie) the week after that.

It will never get in bookworm (that one is frozen for new software), though I could upload a package to bookworm-backports once it is in testing.

**Comment by dsommers** [comment]
@Drizzt321 In parallel to Marc's work, we will also provide some new repositories for our Cloud Connexa users providing only stable releases.  That is undergoing QA right now.  Once that is done, documentation will be updated.  And I will ensure our builds are up to the same standards as the Marc's packaging.

But attached is a v21 .deb package for Debian 12 from our own internal builds: [openvpn3_21-1+bookworm_amd64.deb.gz](https://github.com/OpenVPN/openvpn3-linux/files/12856029/openvpn3_21-1%2Bbookworm_amd64.deb.gz)

**Comment by mleeman** [comment]
took some time to get feedback on Debian, I am now processing it.

**Comment by mleeman** [comment]
https://salsa.debian.org/erpel/openvpn3-client/-/blob/master/debian/patches/0001_openvpn-to-_openvpn-user.patch

**Comment by mleeman** [comment]
I have to remove the unicode-impl.hpp because of the copyright, by pulling and adjusting the one from llvm this seem tackled.

That one has a more recent copyright and should not be a problem for DFSG

**Comment by mleeman** [comment]
As far as I know, the package can be uploaded to Debian, I don't think anything should block the upload. The DFSG patch was from another package that is free (in essence, a newer licence header).

The only thing I need, is a sponsor; I have some ppl I could ask, but I preferred 'to follow the process' over calling in some favour.

There was someone from OpenVPN that said he'd sponsor, but haven't heard from  him after I asked.

**Comment by alee-ntap** [comment]
> As far as I know, the package can be uploaded to Debian, I don't think anything should block the upload. The DFSG patch was from another package that is free (in essence, a newer licence header).

The package will be reject directly from NEW if the original upstream contains any non DFSG code. Please upstream your DFSG-free patch instead. And then roll a git release tarball or wait for a new upstream release for the Debian package upload.

> The only thing I need, is a sponsor; I have some ppl I could ask, but I preferred 'to follow the process' over calling in some favour.
> 
> There was someone from OpenVPN that said he'd sponsor, but haven't heard from him after I asked.

I am surely can find you sponsor once you solved all the issues in the package. The upstream tarball should no contains any non-DFSG license. And fixed all the issue that lintian detected on mentors:
   https://mentors.debian.net/package/openvpn3-client/
Highly suggest to use DEP5 format for the `debian/copyright` file. So that can be parse by montors. 

> I've uploaded it twice already to mentors, but it gets stuck there.

And I also highly recommend to use the source package name matches with upstream tarball unless it has been taken by another package in Debian. This helps you to keep your `debian/watch` file simple. You probably should update the package name in the ITP #904044 as well.

**Comment by mleeman** [comment]
It's been a couple months since I worked on it, but AFAIK

1. The code has been repackaged to exclude the non-dfsg code
2. The watch file is not there because OpenVPN does not provide an indexable web site: there is no way that I know of that I could scan for new releases. I checked this off with upstream, and there was word that it was something that would be considered in the future
3. The update of the header file that has the updated copyright is something that upstream is considering; but it needs to be integrated in an upstream release and when the problem was raised, they were too far in their test cycle.
4. The package name was renamed with a clear purpose: upstream calls it 'openvpn3-linux'. I do not have a problem with the naming; but the release openvpn3-linux is a combination of to parts: a library part and a client part that is released as a single archive. As far as I understood the longer term goals of upstream, the 'library' part is intended to be re-used for a server component in the future. With that reasoning, I concluded that 'openvpn3-client' would reflect the reality better than 'openvpn3-linux', especially when upstream might have an 'openvpn3-server' component and possibly s separate library release component.

It's true, I decided to keep on referencing the single and validated source release of upstream and leave the bundling to them. This is also the reason why I have not split it off at the moment: the library is only used by a single application (i.e.. openvpn3-client).

Picking openvpn3-client as a package is IMHO future proof and avoids creating transitional packages and reflects the reality better: contrary to openvpn2, openvpn3 only has a client component at the moment. `openvpn3-client` over `openvpn3-linux` has the benefit of clarity.

**Comment by dsommers** [comment]
> And I also highly recommend to use the source package name matches with upstream tarball unless it has been taken by another package in Debian. This helps you to keep your debian/watch file simple. You probably should update the package name in the ITP #904044 as well.

Just beware that this is a bit "complicated".

Upstream projects:
* `openvpn3` -> OpenVPN 3 Core Library (library implementing the OpenVPN protocol as a C++ library)
* `openvpn3-linux` -> Linux focused implementation of the OpenVPN 3 Core library, with the full config/session management stack

Package names used in the OpenVPN Inc provided packages:
* `openvpn3` -> Base package, having the core functionality for the OpenVPN 3 Linux service stack and front-end tools
* `openvpn3-client` -> Provides the OpenVPN 3 client components (provides the pieces to connect to a server)
* `openvpn3-selinux` -> Provides SELinux policies for the OpenVPN 3 Linux stack for RPM packages (part of `openvpn3` in .deb)
* `openvpn3-devel` -> Provides C header files for some constant values (part of `openvpn3` in .deb)
* `openvpn3-addon-aws` -> Optional package, provides AWS VPC integration support

In the future, when efforts to provide an OpenVPN 3 based server implementation happens, that is expected to end up in an `openvpn3-server` package.

This was discussed with @mleeman a while back, and he decided to put most of this into the `openvpn3-client` package.  I come from Fedora/RHEL packaging where more and smaller packages with a dedicated functionality is the normal approach, while Debian seems to go the opposite way, wanting as much as possible in a single package.

I won't object to to the Debian packaging policies, I'm sure they have valid reasons for their decisions and if Debian wants more into each package, sure, that's fine.  But documentation wise, it would be good if the "client functionality" would be placed in the `openvpn3-client` package.  That would then be consistent across distributions and easier for users to relate to.

**Comment by alee-ntap** [comment]
> 1. The code has been repackaged to exclude the non-dfsg code

Good to know.

> 2. The watch file is not there because OpenVPN does not provide an indexable web site: there is no way that I know of that I could scan for new releases. I checked this off with upstream, and there was word that it was something that would be considered in the future

Please temporary provide a lintian-override with the reason for `debian-watch-file-is-missing`.

> 3. The update of the header file that has the updated copyright is something that upstream is considering; but it needs to be integrated in an upstream release and when the problem was raised, they were too far in their test cycle.

Please update your patches in DEP-3 format with proper tags and address the reason to have these patches and link with open issue addressed in upstream.

[skip..]
> Picking openvpn3-client as a package is IMHO future proof and avoids creating transitional packages and reflects the reality better: contrary to openvpn2, openvpn3 only has a client component at the moment. `openvpn3-client` over `openvpn3-linux` has the benefit of clarity.

Thank you for explain this. Now I understood why it's named `openvpn3-client`. Please update your ITP with the reason you mentioned here and make the `openvpn3-client` package name `matches` with the package name in Debian `ITP` Bug you want to closed in `debian/changelog`.

BTW, does any reason that the package sets Architecture: `amd64` in `debian/control`? 

Please fix or override lintian errors that detected in your latest upload on mentors and provide `debian/upstream/metadata` file. Also the `pkg-config` in build-depends has been superseded that was detected on mentors.

**Comment by dsommers** [comment]
@alee-ntap 

> BTW, does any reason that the package sets Architecture: amd64 in debian/control?

Yes, there has been some challenges compiling it successfully on 32-bit CPU architectures.  There are some challenges with the C++ templates ending up with the wrong and conflicting data types; iirc it was related to 32-bit vs 64-bit `int` and `long` data types.

This might change with the v22_dev work, where there is a brand new D-Bus integration where some if this code is changed and done a bit differently.  But I'm not ready yet to lift the 32-bit exclusion.  It needs to be thoroughly tested.

**Comment by alee-ntap** [comment]
> @alee-ntap
> Yes, there has been some challenges compiling it successfully on 32-bit CPU architectures. There are some challenges with the C++ templates ending up with the wrong and conflicting data types; iirc it was related to 32-bit vs 64-bit `int` and `long` data types.

In that case, we can specific more 64-bit architectures that Debian 13 will supported, eg:
```
Architecture: amd64 arm64 mips64el ppc64el riscv64 s390x
```

**Comment by mleeman** [comment]
I did a test build locally with arm64 and that one seems to compile, we just assume that that the architectures above work too and let the debian builders veirfy them (takes me the better part of an hour to verify a architecture with muti-arch/pdebuild)/cowbuilder.

**Comment by dsommers** [comment]
The Fedora Copr builds we already do a similar set of architectures as @alee-ntap suggests.  They all build fine, but I don't know if anyone uses it on those.  We do have reports though that the arm64 works on the 64-bit capable RPi.

**Comment by alee-ntap** [comment]
> I did a test build locally with arm64 and that one seems to compile, we just assume that that the architectures above work too and let the debian builders veirfy them (takes me the better part of an hour to verify a architecture with muti-arch/pdebuild)/cowbuilder.

I already tested it works in an arm64 VM. Let's get all the lintian error fixed and upload into Debian. Would you prefer to use mentors or just your package git repo on salsa for reviews?

**Comment by mleeman** [comment]
> I already tested it works in an arm64 VM. Let's get all the lintian error fixed and upload into Debian. Would you prefer to use mentors or just your package git repo on salsa for reviews?

I don't mind where the reviews are as long I as I pick them up, but a detailed Debian specific review might be better off on salsa.

I uploaded a new copy, the watch warning is still there (even with the `debian/source/lintian-overrides`), added DEP3  headers (tnx, need to do that for another project too), DEP5 is on the back burner.

**Comment by alee-ntap** [comment]
> I uploaded a new copy, the watch warning is still there (even with the `debian/source/lintian-overrides`), added DEP3 headers (tnx, need to do that for another project too), DEP5 is on the back burner.

I saw it's override correctly in your new upload:
![Screenshot 2024-02-20 at 2 32 19 PM](https://github.com/OpenVPN/openvpn3-linux/assets/142503442/b5c02c77-17dc-4488-8e9e-14b5e67dddef)

Now only `redundant-control-relation` that you have two entries for `libxml2-utils`. That can be easily fixed by `wrap-and-sort` command from `devscripts` package.

Besides the lintian errors. Why does the `debian/rules` file has to specify these:
```
$(shell dpkg-buildflags --export=configure) --host=$(DEB_HOST_GNU_TYPE) --build=$(DEB_BUILD_GNU_TYPE)
```
```
--disable-unit-tests \
--disable-build-test-progs \
```
[skip..]
```
$(EXTRA_ARGS)
```
?

**Comment by dsommers** [comment]
I would recommend to keep `--disable-build-test-progs` ... it avoids building lots of development test programs, not installed (and not really useful for end-users).  That speeds up the compiling a bit.  The unit-tests can be valuable as a check-point, if they are run later on in the build/packaging process.

**Comment by mleeman** [comment]
I either based my work on openvpn2 packaging or used work from @dsommers , I don't remember. In any case, I'll enable the unit testing (2 tests I see) and remove the unused variable.

```
make[5]: Entering directory '/home/marc/Development/salsa/openvpn3-client'
PASS: src/tests/unit/unit-tests
PASS: src/policy/syntax-check.sh
============================================================================
Testsuite summary for OpenVPN3/Linux 21
============================================================================
# TOTAL: 2
# PASS:  2
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================

```

The test-progs we leave out?

**Comment by alee-ntap** [comment]
> I either based my work on openvpn2 packaging or used work from @dsommers , I don't remember. In any case, I'll enable the unit testing (2 tests I see) and remove the unused variable.

It's better to have the review on your git repo on salsa then. :D

> The test-progs we leave out?

I think so. Unless it would be useful to provide in separate binary package eg: `openvpn3-client-test-progs`?

**Comment by dsommers** [comment]
> I think so. Unless it would be useful to provide in separate binary package eg: openvpn3-client-test-progs?

There might be one or two programs here which might be handy in really low level debugging.  But if you need this kind of debugging, then you most likely are already deep into the code itself - so you have these tools easily available in a local build.  Most of these programs are more to test various internal APIs and was more useful in the early days of the implementation before all the APIs had settled.  If any of these tests programs are useful for general testing, I tend to convert them into a unit-tests instead.  I would just leave them aside for now.

In regards to the unit-test being disabled; yeah, we might have disabled that in the OpenVPN Inc packaging, as some of these tests depends on D-Bus and some systemd capabilities being available and functional.  We do these builds inside docker containers where that is not the case, so the unit test would fail due to infrastructure issues.  These tests are run elsewhere in our build infrastructure, as well as I'm running them regularly during development.

**Comment by alee-ntap** [comment]
@mleeman Great! You have mostly fixed all the issue that lintien detects. \o/
Let's fix the `debian/copyright` file. You made mistake in `Source:` field and I think  you should add `Comment:` field as the source tarball has modified due to DFSG reason. More info for the fields you can find here:
https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/

I found more than one Authors in the `git log`. Please add them into the copyright list. And also add yourself and if any others into `Files: debian/*`.

Here are an example package that also `License: AGPL-3+`:
https://salsa.debian.org/java-team/pdfsam/-/blob/master/debian/copyright

And I highly recommend to use `licensecheck` to review the copyright,. I did that and surprisedly found there are a lot of info are missing in your `debian/copyright` file. You can find more details for copyright review tools here:
https://wiki.debian.org/CopyrightReviewTools

**Comment by alee-ntap** [comment]
> ```
> $(shell dpkg-buildflags --export=configure) --host=$(DEB_HOST_GNU_TYPE) --build=$(DEB_BUILD_GNU_TYPE)
> ```
I think this can be drop as Debhelper compatibility level 13 already handles that.

**Comment by mleeman** [comment]
Thanks for the review @alee-ntap , very much appreciated.

The licence is something that I need to get a hang of, the debian/* part was removed because someone else suggested me that it was not needed if I kept the debian/* directory under the same licence as upstream.

```
Files: *
Copyright: 2010-2023, OpenVPN Solutions LLC
License: AGPL-3

Files: debian/*
Copyright: 2023-2024, Marc Leeman <marc.leeman@gmail.com>
License: AGPL-3
```

The only thing that would warrant a difference here is the dates because the upstream code was released in 2023, while we are still working on the rest.

In `git log` (salsa), I only see one name (mine), though the git commit might show a different e-mail. That should not matter, I think.

I am certain a lot of ppl have worked on the openvpn3 code itself; but shouldn't that be covered by the over arching copyright when released by upstream: if they state AGPL; does packaging need to provide more information?

The licensecheck is a nice tool, and I see indeed that some copyrights are missing in debian/copyright (.e.g autotools/generated files). Again, do we need to provide the licence information that is part of upstream (and in this case, generated files)? I cannot imagine that this package is the only one that would sin against this.

**Comment by alee-ntap** [comment]
> The only thing that would warrant a difference here is the dates because the upstream code was released in 2023, while we are still working on the rest.

Right, so keep it as same as the example package will helps. :)

> In `git log` (salsa), I only see one name (mine), though the git commit might show a different e-mail. That should not matter, I think.
There is not need to list multiple emails with the same person.

> I am certain a lot of ppl have worked on the openvpn3 code itself; but shouldn't that be covered by the over arching copyright when released by upstream: if they state AGPL and covered with their company; does packaging need to provide more information, by extracting the authors from git?

Correct, as the ftp-master will check to ensure the copyright is completely listed and without any non-DFSG code and binary file. Or your package will get rejected. That will takes more time to get another sponsor upload.

> The licensecheck is a nice tool, and I see indeed that some copyrights are missing in debian/copyright (.e.g autotools/generated files). Again, do we need to provide the licence information that is part of upstream (and in this case, generated files)? I cannot imagine that this package is the only one that would sin against this.

All the packages in Debian are required to list complete copyright due to Debian policy. And I found file in the `src/tests/` has copyright info are not listed. These are must listed in the copyright file. Also I haven't checked the binaries files, eg: .pdf and .png also needs to check and list the licenses.

I can easily find some examples on
    https://codesearch.debian.net
You may search eg:
```
.png path:debian/copyright
.pdf path:debian/copyright
```

Hope that helpful.

**Comment by mleeman** [comment]
Back to the drawing board, some more work:

```
Hi

There are two license inompatibilities:
(A)GPL-3 and BSD-4-Clause-UC are incompatible (yes, I saw the comment,
but this makes the license either a custom one or possibly a
BSD-3-Clause)
GPL-2 and AGPL-3 are incompatible

AGPL is not suitable for non-network code, so at least debian/* should
get something else.

Your lintian overrides don't match what ftp-master still expects, see
all the errors.

The polkit maintainer just asked to reject new packages for shipping
legacy polkit pkla files.  So please remove the stuff from
/var/lib/polkit-1 and convert to the new-style rules.

Please drop all autoconf/automake generated and vendored files.  We've
just seen what this produces as problems.

There is vendored code, but also a dependency on the same library
(libasio-dev).

Some other remarks:
- Please use dh_installtmpfiles (rename debian/openvpn3-client.conf to
  .tmpfiles I think), enabled in debhelper compat 13
- You could use debhelper compat 14 directly and avoid some of the
  overrides in debian/rules, as it calls dh_installsysusers
- Preemptive architecture limits are not the way to go, please remove
  this list
- It contains DSA keys in /etc/openvpn3/awscerts, which a current crypto
  policy should not really accept
- There seems to be two distinct was to start it,
  openvpn3-autoload.service and openvpn3-session@.service, is this
  necessary?
- /var is not really for user config, but for state.
- Is /var/lib/openvpn3 supposed to be in the package or created via
  systemd-tmpfiles?  Currently I see both.
  ```

**Comment by schwabe** [comment]
Why is AGPL not suitable for non-network code? Sure the difference between AGPL and GPL only affects code that uses network but that does not make AGPL3 unsuitable for other code.

**Comment by mleeman** [comment]
@schwabe TBH, I do not know. One thing I've learnt when packaging this package; every different DD will give you different answers, especially for this software that has so many contribution from different corners.

I think, in this particular case, is to change the licence of the debian directory.

I  have had a quick look and some other remarks are more technical so I can start working on those.

**Comment by mleeman** [comment]
It might be quiet about the topic of getting it in Debian, but sometimes I have to find some time to make the changes (or understand them) and getting things moving within Debian is another matter too; it just takes more time than anticipated.

I cannot remember that this was the case years ago...

**Comment by schwabe** [comment]
@mleeman to be honest that debian directory cannot be AGPL3 if the whole project is AGPL3 sounds ridicioulous and probalby been done by someone with little experience/knowledge in liceses. That would like saying, you have to put man pages under a different license in GPL projects since GPL has section about code and libraries and man pages are only text.

**Comment by mleeman** [comment]
FYI:

There are some minor spelling errors that lintian picked up:
https://salsa.debian.org/erpel/openvpn3-client/-/blob/master/debian/patches/0003_lintian-spelling-errors.patch?ref_type=heads

**Comment by Phhere** [comment]
Is there any plan to support debian unstable / trixie? currently it is not possible to install it there

**Comment by mleeman** [comment]
yes there is, 

There have been a number of iterations that improved the quality of the package itself and the licensing was also a bit of a struggle to get through (and it needed some review).

Unfortunately; Debian is not what is used to be wrt the inclusion of new packages: I have uploaded the package twice to be included via mentors only to be removed again after a timeout since it no-one was willing to sponsor it.

Coincidentally, I was planning to upload the package again with hopes that it would be picked up at some point.

**Comment by detlefla** [comment]
Is your package available somewhere (as a .deb)?
Building from source fails for me (e.g. libgdbuspp-dev cannot be found).

**Comment by detlefla** [comment]
Thank you for making the package available. It doesn't install, however, on testing / unstable (trixie) as it depends on libtinyxml2-9 and trixie has libtinyxml2-10. (That's the one it fails on; not sure if there are more broken dependencies.)

**Comment by mleeman** [comment]
I will make a package available tomorrow for unstable.

**Comment by alee-ntap** [comment]
@mleeman Would you mind to forward previous reject message from Debian's ftp-master to your ITP on Debian's bug tracker, and continue to work to make it in Debian unstable? I know there are some reasons you possibility doesn't agree. So make it in the public in the ITP and request for help then.

**Comment by mleeman** [comment]
> @mleeman Would you mind to forward previous reject message from Debian's ftp-master to your ITP on Debian's bug tracker, and continue to work to make it in Debian unstable? I know there are some reasons you possibility doesn't agree. So make it in the public in the ITP and request for help then.

Yeah, I'll pick it up again: I gave it a breather because of the long discussions about the licence checks. Unrelated, I had a reject on a introducing a package because it was circumventing bandwidth limitations :-) (`apt-fast`).

**Comment by mleeman** [comment]
@alee-ntap The feedback of ftp-master was handled IIRC; I prepared a package (I believe just before the summer) and got it uploaded and reviewed on mentors. There was even an endorsement of a very helpful British guy; but no-one was interested in picking it up for for upload.

Like I said, I'll push it again and see if I can get things rolling again...

**Comment by detlefla** [comment]
@mleeman Many thanks for providing the .deb! It's installable and runnable. Can't make it connect to our server yet – but this may be a local problem.

**Comment by alee-ntap** [comment]
@mleeman 
> @alee-ntap The feedback of ftp-master was handled IIRC; I prepared a package (I believe just before the summer) and got it uploaded and reviewed on mentors. There was even an endorsement of a very helpful British guy; but no-one was interested in picking it up for for upload.
> 
I uploaded it once for you into NEW queue, but got rejected by ftp-master. 

> Like I said, I'll push it again and see if I can get things rolling again...
And I can upload it again. It would be nice to append the rejected message into your ITP. So that we can check if all the issue mentioned by ftp-master got fixed in the next try. :)

**Comment by mleeman** [comment]
@alee-ntap 

Sorry for the delay, was caught up i n other work. Unfortunately, I fail to find the  original mail from ftp-master, but the ITP [1] has remarks and replies from a couple of people, so has the discussion on mentors [2] and the RFS [3]. I uploaded a new version last week that disables the aws service at the moment (default it failed to start so it's good to get that out).

I believe that most of the remarks that can be addressed, have been addressed.

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=904044
[2] https://mentors.debian.net/package/openvpn3-client/
[3] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1053565

**Comment by danterolle** [comment]
Hello everyone! I'm joining those who have tried building openvpn3-linux. In the guide, [libssl1.1 is listed among the dependencies](https://github.com/OpenVPN/openvpn3-linux/blob/master/BUILD.md#debianubuntu), but if I'm not mistaken, it is no longer needed (on Debian 12, for instance, the package is not found). Perhaps it should be removed? I'm eagerly waiting for libgdbuspp-dev to become available, and especially for the .deb package.

**Comment by mleeman** [comment]
@danterolle if you use the sources on salsa [1], you can just use the standard debian tooling; I've built it in a clean chroot a couple weeks ago for bookworm and trixie.

libssl1.1 is a dependency that is generated during packaging. See the posts above for packages built for debian 12 and debian 13.

[1] https://salsa.debian.org/erpel/openvpn3-client

**Comment by dsommers** [comment]
@mleeman I just had a look at your salsa tree .... that is quite outdated now, unless I looked at the wrong places.

The v23 release is the latest stable release (I'm wrapping up v24 these days).  One huge and one important difference ... 

* autotools are now removed in favour of the Meson build system (that's the huge change)
* You also need the GDBus++ library now to build anything newer than the `v21` release.  Everything needed is here: https://codeberg.org/OpenVPN/gdbuspp/

I believe I have applied some suggested changes from you, but I'm not up-to-date right now if anything is missing.  Please bug me if there are things which could be done better - and I'll try to get things sorted out quicker.

**Comment by mleeman** [comment]
No, no, it is still the same one.

bummer, I must have missed the releases, I'll update the tree.

I'll need to have a look at the changes, but the most important ones for me at the moment are listed under DFSG changes:
https://salsa.debian.org/erpel/openvpn3-client/-/blob/master/debian/README.source?ref_type=heads

These require me to recreate the release tarball.

**Comment by dsommers** [comment]
I'll follow up on the Sparc binary ... I have a few options here.  It's too late for the v24 release (I pushed out an updated master the yesterday).

In regards to the unicode headers ... the LLVM versions did not work out in the Core library without more modifications in the Core library.  This is on our todo list to fix.  But it's a different project OpenVPN 3 Linux depends on, so it's a bit more work to complete this one.

When it comes to a "watch file" for releases ... this should be possible to use: https://swupdate.openvpn.net/community/releases/openvpn3-linux.latest

**Comment by mleeman** [comment]
I am having a look at updating the the release of September. The most important issue atm is that the new version depends on a new library that is also not in Debian. Packaging is not a problem (done it already), but getting it in will take some time (ITP, RFS, review, ...).

Only then will I be able to upload the new openvpn3-client.

I'll keep on trying to get the v21 in in parallel since this one has been reviewed extensively.

I can provide the working progress packages as above.

The gdbuspp packages build for bookworm and trixie, so that's good.

```
'../libgdbuspp3_2~televic12+1_amd64.deb'.
'../libgdbuspp3-dbgsym_2~televic12+1_amd64.deb'.
'../libgdbuspp-dev_2~televic12+1_amd64.deb'.
```

**Comment by dsommers** [comment]
@mleeman Just a heads-up on v21 and older versions.  They will have issues with glib2-2.76 and newer (see #171 for an extensive discussion).  That's what led to writing the GDBus++ library.

**Comment by dsommers** [comment]
Also ... v24 is being pushed out the door ... our own apt/dnf/yum repos are not yet updated (that's what's brewing right now), but source tarballs are published:

* OpenVPN 3 Linux v24
  <https://swupdate.openvpn.net/community/releases/openvpn3-linux-24.tar.xz>
  <https://swupdate.openvpn.net/community/releases/openvpn3-linux-24.tar.xz.asc>

* GDBus++ v3
  <https://swupdate.openvpn.net/community/releases/gdbuspp-3.tar.xz>
  <https://swupdate.openvpn.net/community/releases/gdbuspp-3.tar.xz.asc>

**Comment by mleeman** [comment]
No worries, I'm working on getting this in Debian for a year now, not about to give up now...

**Comment by dsommers** [comment]
@mleeman Just for clarity ... Those sources I pointed at are the official release source tarballs; they are complete.  I've just not announced the release yet.  We have our own set of package repos in parallel to upstream Linux distro repos, the announcement will be done when those repos are updated.

Here is a signed blob (with the security@openvpn.net key, same key used in the .asc files above) with the SHA256 hashes of all the files:
```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

9ecf8dccdbc601c4325b0248db7cb1e39c8689e3b99f5fc801b42056d68a7256  openvpn3-linux-24.tar.xz
a3d6bd735d46958f2458484a4338eaf894e710ac895852c9c734671a2e46e821  openvpn3-linux-24.tar.xz.asc
c7a053a13c4eb5811a542b747d5fcdb3a8e58a4a42c7237cc5e2e2ca72e0c94e  gdbuspp-3.tar.xz
b9cf732d7a347f324d6a5532dc48f80c2815dbf6704c169b4ee97a411506a99b  gdbuspp-3.tar.xz.asc
-----BEGIN PGP SIGNATURE-----

iQJJBAEBCAAzFiEEV+lQSddqo5p0Q5YDUzpoYFKfI8UFAmdRZBwVHHNlY3VyaXR5
QG9wZW52cG4ubmV0AAoJEFM6aGBSnyPF9VUP/jXBW/EtQ4Efgz87mGklOOtMGYax
nz5Dt8KOAFuSj38y7Y1Zwo5cQQlnVcZGmfByohalf3XDGpb3Pci2cIvB42cSZYh3
z7oSZvCorndKwmhQIByz0FSiLwD0iQ4gEwdPO7UjES7Vb3X9iYU+6ijQkvfuaYfe
UQ7Vmc8+lHDNp0iqepM9mw2dDGOx4A03l8nsvxvgWLIiZ1/1/IVBmNaifZEpQH5S
KOOCElkSlQT+zZ9rgKtuLmPw2hJhsLIDmdn/yobdpZ2wR2pWxPCg8HxUFEi9AvoZ
+K2uc8KPh+04YNfo8RJKpdveAtMETpwF+dRiTk3lPYdfD2horsIk89hgFpXi2UPU
krE0u68woUAJFMy2tWFR4IrQBFptT0Mp14R4ddkJId2ukemVuhwlimdv61uzJx9r
kYfyc3H66sPKcexDnVSFi+aJEzo7DjpUx4oq4Ix64DXVBRqmCzGnc8jEiUOn5MdB
PMi7ipLHJiSqfAMWhNC424Oudrv4/ytXL8j7bamj2bl6tZW0uYc4tCrAA/37kwB5
Aud8gW4w0gZVY6rA5tOYSjikIndQu8f0UFBsoryjUNJBBJNDw8tcDZY9YeWIkTdi
glLp6RuHnk4dLkgOARK9qARKefFscx/VSnPxnWBjfW73vBllz3mXX0CSFmDgS5UB
EKHEgnllibySz+t+
=RPsC
-----END PGP SIGNATURE-----
```

**Comment by mleeman** [comment]
@dsommers 

repositories are up to date, packages build in clean chroot. `openvpn3-client` is still dfsg

https://salsa.debian.org/erpel/openvpn3-client
https://salsa.debian.org/erpel/gdbuspp

once I am at home, my ITP will be sent by my MTA; when I have some time, I'll upload/update the versions on my OBS repositories too.

**Comment by dsommers** [comment]
@mleeman whoops!  That's an ugly mishap!  I agree, that patch should fix it.

Could you send a patch to the openvpn-devel mailing list with your fix?  Then I'll get that applied to git master asap.

**Comment by mleeman** [comment]
after I get my VPN connection up and running ;-) 

getting this with 24, investigating

```
Dec 06 09:21:43 moya net.openvpn.v3.log[1403]:  {tag:13961402122662001886} Client !! CRITICAL !!: Failed changing DNS Scope: No DNS resolver configured
```

**Comment by dsommers** [comment]
@mleeman Stop all the `openvpn3-service-*` processes (`kill -INT` will be enough).  Then try running `openvpn3-admin init-config --write-configs` as `root`.  See if that helps.

In our own .deb packaging we do this in a postinst script:
```
# Do not fail, just log any errors
set +e
LOGDEST="/var/lib/openvpn3/openvpn3-init-config.log"
echo "" >> "$LOGDEST"
echo "** openvpn3-admin init-config start -- `date`" >> "$LOGDEST"
/usr/sbin/openvpn3-admin init-config --write-configs >> "$LOGDEST" 2>&1
echo "** openvpn3-admin init-config done (exit-code: $?)" >> "$LOGDEST"
```

The `--write-configs` can be removed for a "dry-run".  And it will never overwrite any pre-existing configuration files, unless you give the `--force` argument in addition.

**Comment by mleeman** [comment]
Thanks, that did the trick, I'll add it.

**Comment by dsommers** [comment]
@mleeman FYI ... I've applied your patch and fixed spelling issues.  They're in the  `master` branch, commit 9a1cf3fae9fb3788e9714d148d9b7efcb5f4c948 and commit 6b16d119810b0cec66fab1afa79c33ecb6c73234

**Comment by alee-ntap** [comment]
@mleeman Sure.
Forgot to push them to into your repo on salsa?
    https://salsa.debian.org/erpel/openvpn3-client

**Comment by mleeman** [comment]
done, you're right. Is normally part of gbp tag; gbp push, but as long as it is not accepted, I'm not tagging it.

there is a second one with the bindings now, there is no RFS for that one since it's tied in with the client.

1. https://salsa.debian.org/erpel/openvpn3-client
2. https://salsa.debian.org/erpel/gdbuspp

tnx.

**Comment by dsommers** [comment]
It's fine to use the codeberg.org link for this project.

**Comment by mleeman** [comment]
I'll update it in git/salsa.

**Comment by mleeman** [comment]
Good news everyone. 

We finally managed to get openvpn3-client into Debian. It was not the most simple software to package (licence related and embedded 3rd party code that had to be removed at the 11th hour), but we made it.

It will now be part of the upcoming Debian 13 release.

A special tanks to @dsommers for the upstream help and support and to @alee-cd for the review and sponsorship and Phil Wyett for review and licence help..


![Image](https://github.com/user-attachments/assets/9ab6f7bc-6cc1-42b6-9e93-3f549cd6ff52)

**Comment by mleeman** [comment]
@FabioPedretti 

Yes, gdbuspp was in Debian for a bit longer (we got the dependency in before submitting the package itself, it avoided discussions). At the moment (I think we discussed this at some point in this (tldr;) thread, the package only compiles correctly on 64bit architectures. Initially, I limited the architecture list; but got slapped on the wrist about that (this is not the way to go).

I will change the homepage in gdbuspp.

**Comment by dsommers** [comment]
@FabioPedretti That's fixed in git master.  That fix will arrive in v25, but you can cherry-pick commit b87a077c2ef2eafc12f744e780cdb23a6924e3d2 for your build.

**Comment by dsommers** [comment]
One more note ... I don't know if this is only in your build @FabioPedretti .... but `-Dopenvpn3_core_version=3` is being set in your build.  Please don't do that.  This sets the `OPENVPN_VERSION` macro, which should already have the proper value from `src/build-version.h` in the tarball.  If it's built from a git checkout, it should be able to extract that version string from `openvpn3-core/`.

It is crucial this string matches the OpenVPN 3 Core Library (`openvpn3-core/`) version.  Especially when connecting to OpenVPN Access Server or the Cloud Connexa, this string may result in failed connections.  This version is reported to the VPN server and the server side may do some version validations.

**Comment by dsommers** [comment]
Thank you so much for all your work on this, @mleeman!  I really appreciate all your efforts and handling all the challenges along the way.

**Comment by mleeman** [comment]
I merged it but there is something off. When that variable is not set, it tries to run a script that accesses git resources (making the package build fail).

**Comment by mleeman** [comment]
This is the problem:

```
166 if fs.exists(meson.project_source_root() / '.git')
167     openvpn3_core_ver = get_option('openvpn3_core_version')
168     if openvpn3_core_ver == ''
169        openvpn3_core_ver = run_command(openvpn3_core_inc / 'scripts/version', check: true).stdout().strip()
170     endif
...
```

resulting in 

```
Running command: /home/marc/Development/salsa/openvpn3-client/./openvpn3-core/scripts/version
--- stdout ---

--- stderr ---
fatal: not a git repository: '/home/marc/Development/salsa/openvpn3-client/./openvpn3-core/scripts/../.git'
fatal: not a git repository: '/home/marc/Development/salsa/openvpn3-client/./openvpn3-core/scripts/../.git'



../meson.build:169:27: ERROR: Command `/home/marc/Development/salsa/openvpn3-client/./openvpn3-core/scripts/version` failed with status 128.
```

**Comment by mleeman** [comment]
I think it can be removed no, the value is already filled out in:

`src/build-version.h`

```
/* OpenVPN 3 Core Library version */
#define OPENVPN_VERSION "v3.10.4"
```

```
marc@jenek:~/Development/salsa/openvpn3-client$ cat debian/patches/0004_remove-git-access.patch 
Index: openvpn3-client/meson.build
===================================================================
--- openvpn3-client.orig/meson.build
+++ openvpn3-client/meson.build
@@ -159,30 +159,8 @@ if (dep_tinyxml2.version().version_compa
     endif
 endif
 
-# Prepare some version details - if on a git checkout
-# For tarball builds, the build-version.h should already pre-exist
-# in ./src
 fs = import('fs')
-if fs.exists(meson.project_source_root() / '.git')
-    openvpn3_core_ver = get_option('openvpn3_core_version')
-    if openvpn3_core_ver == ''
-       openvpn3_core_ver = run_command(openvpn3_core_inc / 'scripts/version', check: true).stdout().strip()
-    endif
 
-    version_header = configuration_data()
-    version_header.set_quoted('OPENVPN_VERSION',
-                              openvpn3_core_ver,
-                              description: 'OpenVPN 3 Core Library version')
-    version_header.set_quoted('PACKAGE_GUIVERSION',
-                              run_command(['scripts/get-version','--gui-version'], check: true).stdout().strip(),
-                              description: 'OpenVPN 3 Linux version identifier')
-    version_header.set_quoted('PACKAGE_NAME', 'OpenVPN3/Linux',
-                              description: 'Name of this project')
-    configure_file(
-        configuration: version_header,
-        output: 'build-version.h'
-    )
-endif
 meson.add_dist_script(find_program('scripts/prepare-build-version-h'))
 cp_prog = find_program('cp', required: true)
```

is building.

**Comment by mleeman** [comment]
I have pushed the change to master, it will be part of the next upload unless something unexpected pops up. Thanks for the contribution and suggestions @FabioPedretti

**Comment by dsommers** [comment]
I'm closing this ticket now.  Debian 13 will have the OpenVPN 3 Linux packages from @mleeman.  And I've seen the same packaging appearing in Ubuntu 25.04.

A huge THANK YOU for excellent effort goes to @mleeman!
